datumaro/plugins/synthetic_data/image_generator.py

```diff
@@ -0,0 +1,225 @@
+        pixels = -1
+        i = 0
+        while pixels < self._threshold and i < self._iterations:
+            param_size = rng.randint(2, 7)
+            params = np.zeros((param_size, IFSFunction.NUM_PARAMS + 1), dtype=np.float32)
+
+            sum_proba = 1e-5
+            for p_idx in range(param_size):
+                a, b, c, d, e, f = [rng.uniform(-1.0, 1.0) for _ in range(IFSFunction.NUM_PARAMS)]
+                prob = abs(a * d - b * c)
+                sum_proba += prob
+                params[p_idx] = a, b, c, d, e, f, prob
+            params[:, IFSFunction.NUM_PARAMS] /= sum_proba
+
+            fractal_img = self._generate_image(rng, params, self._num_of_points, base_h, base_w)
+            pixels = np.count_nonzero(fractal_img) / (base_h * base_w)
+            i += 1
+        return params
+
+    def _initialize_params(self) -> None:
+        if self._count < self._weights.shape[0]:
+            self._weights = self._weights[: self._count, :]
+
+        instances_categories = np.ceil(self._count / self._weights.shape[0])
+        self._instances = np.ceil(np.sqrt(instances_categories)).astype(int)
+        self._categories = np.ceil(instances_categories / self._instances).astype(int)
+
+    @staticmethod
+    def _create_weights(num_params):
+        # weights from https://openaccess.thecvf.com/content/ACCV2020/papers/Kataoka_Pre-training_without_Natural_Images_ACCV_2020_paper.pdf
+        BASE_WEIGHTS = np.ones((num_params,))
+        WEIGHT_INTERVAL = 0.4
+        INTERVAL_MULTIPLIERS = (-2, -1, 1, 2)
+        weight_vectors = [BASE_WEIGHTS]
+
+        for weight_index in range(num_params):
+            for multiplier in INTERVAL_MULTIPLIERS:
+                modified_weights = BASE_WEIGHTS.copy()
+                modified_weights[weight_index] += multiplier * WEIGHT_INTERVAL
+                weight_vectors.append(modified_weights)
+        weights = np.array(weight_vectors)
+        return weights
+
+    @classmethod
+    def _download_colorization_model(cls, path: str) -> None:
+        proto_file_name = cls._MODEL_PROTO_NAME
+        config_file_name = cls._MODEL_CONFIG_NAME
+        hull_file_name = cls._HULL_PTS_FILE_NAME
+        proto_path = osp.join(path, proto_file_name)
+        config_path = osp.join(path, config_file_name)
+        hull_path = osp.join(path, hull_file_name)
+
+        if not (
+            osp.exists(proto_path) and osp.exists(config_path) and osp.exists(hull_path)
+        ) and not os.access(path, os.W_OK):
+            raise ValueError(
+                "Please provide a path to a colorization model directory or "
+                "a path to a writable directory to download the model"
+            )
+
+        def _download_file(url, output_path, timeout=60):
+            response = requests.get(url, timeout=timeout)
+            response.raise_for_status()
+            with open(output_path, "wb") as fd:
+                for chunk in response.iter_content(chunk_size=128):
```

Comment:
1. 128 seems like a really small chunk size. I would suggest using a million or so.
2. Chunked downloading introduces its own problem: if the download is interrupted, then a partial file will be left on disk, which means that the next time the image generator is used, it will try to load the partial file and fail. To fix this, the check for file existence must verify not just the existence, but the length too (verifying the hash digest would be ideal, but maybe that's too slow).

Reply:
3. For chunked downloading to be more efficient than fetching the entire file at once, you need to add `stream=True` to the `get` call, and for _that_ to be exception-safe, you need to use `response` as a context manager.

Reply:
Fixed

Reply:
1. The chunk size is set to 1 mb
2. File removal is added
3. stream is added

Reply:
2. File removal is nice, but it is insufficiently robust. The Datumaro process could get killed in the middle of downloading without getting a chance to clean up, and the next time it is started, it will load a truncated file.

  Thinking about it again, there's actually a simpler way of ensuring that only a complete file is loaded. You first download to a file with a `.tmp` suffix, and then, after all checks are successful, rename it to remove the suffix.

3. But you didn't ensure that the response is closed by the end.